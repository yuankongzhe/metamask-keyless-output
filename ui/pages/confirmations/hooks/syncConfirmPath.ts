import { useEffect } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import {
  CONFIRM_TRANSACTION_ROUTE,
  SIGNATURE_REQUEST_PATH,
} from '../../../helpers/constants/routes';
import { currentConfirmationSelector } from '../../../selectors/confirm';

const syncConfirmPath = () => {
  const history = useHistory();
  const { id: paramsTransactionId } = useParams<{ id: string }>();

  const currentConfirmation = useSelector(currentConfirmationSelector);

  // Redirect below is done to keep the confirmation routes backward compatible
  useEffect(() => {
    if (!currentConfirmation) {
      return;
    }
    if (paramsTransactionId !== currentConfirmation.id) {
      const isSignature = Boolean(currentConfirmation.msgParams);
      if (isSignature) {
        history.replace(
          `${CONFIRM_TRANSACTION_ROUTE}/${currentConfirmation.id}/${SIGNATURE_REQUEST_PATH}`,
        );
      }
      // todo: logic to replace url for rest of transactions to be added here
    }
  }, [currentConfirmation, paramsTransactionId]);
};

export default syncConfirmPath;
