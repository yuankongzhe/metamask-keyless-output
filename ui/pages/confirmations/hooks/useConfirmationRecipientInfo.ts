import { useSelector } from 'react-redux';

import {
  accountsWithSendEtherInfoSelector,
  currentConfirmationSelector,
} from '../../../selectors';
import { getAccountByAddress } from '../../../helpers/utils/util';

function useConfirmationRecipientInfo() {
  const currentConfirmation = useSelector(currentConfirmationSelector);
  const allAccounts = useSelector(accountsWithSendEtherInfoSelector);

  let recipientAddress = '';
  let recipientName = '';

  if (currentConfirmation) {
    const { msgParams } = currentConfirmation;
    // url for all signature requests
    if (msgParams) {
      recipientAddress = msgParams.from;
      const fromAccount = getAccountByAddress(allAccounts, recipientAddress);
      recipientName = fromAccount?.metadata?.name;
    }
    // TODO: as we add support for more transaction code to find recipient address for different
    // transaction types will come here
  }

  return {
    recipientAddress,
    recipientName,
  };
}

export default useConfirmationRecipientInfo;
