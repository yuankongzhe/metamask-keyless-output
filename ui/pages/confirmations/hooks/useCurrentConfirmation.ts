import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import { ApprovalRequest } from '@metamask/approval-controller';
import { Json } from '@metamask/utils';

import {
  latestPendingConfirmationSelector,
  pendingConfirmationsSelector,
  unapprovedPersonalMsgsSelector,
} from '../../../selectors';

type Approval = ApprovalRequest<Record<string, Json>>;

const useCurrentConfirmation = () => {
  const { id: paramsTransactionId } = useParams<{ id: string }>();
  const unapprovedPersonalMsgs = useSelector(unapprovedPersonalMsgsSelector);
  const latestPendingConfirmation: Approval = useSelector(
    latestPendingConfirmationSelector,
  );
  const pendingConfirmations: Approval[] = useSelector(
    pendingConfirmationsSelector,
  );
  const [currentConfirmation, setCurrentConfirmation] =
    useState<Record<string, unknown>>();

  useEffect(() => {
    let pendingConfirmation: Approval | undefined;
    if (paramsTransactionId) {
      if (paramsTransactionId === currentConfirmation?.id) {
        return;
      }
      pendingConfirmation = pendingConfirmations.find(
        ({ id: confirmId }) => confirmId === paramsTransactionId,
      );
    } else {
      pendingConfirmation = latestPendingConfirmation;
    }
    if (
      pendingConfirmation &&
      pendingConfirmation.id !== currentConfirmation?.id
    ) {
      // currently re-design is enabled only for personal signatures
      // condition below can be changed as we enable it for other transactions also
      const unapprovedMsg = unapprovedPersonalMsgs[pendingConfirmation.id];
      if (!unapprovedMsg) {
        return;
      }
      const { siwe } = unapprovedMsg.msgParams;

      if (!siwe?.isSIWEMessage) {
        setCurrentConfirmation(unapprovedMsg);
      }
    }
  }, [latestPendingConfirmation, paramsTransactionId, unapprovedPersonalMsgs]);

  return { currentConfirmation };
};

export default useCurrentConfirmation;
