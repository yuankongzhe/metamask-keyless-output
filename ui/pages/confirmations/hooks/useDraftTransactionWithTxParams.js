import { useSelector } from 'react-redux';
import { getCurrentDraftTransaction } from '../../../ducks/send';
import { getUnapprovedTransactions } from '../../../selectors';

/**
 * Returns an object that resembles the txData.txParams from the Transactions state.
 * While processing gas details for send transaction and edit transaction,
 * the gas data from draftTransaction and unapprovedTx has to be reorganized
 * to mimic the txdata.txParam from a confirmTransaction
 *
 * @returns {object} The transaction data
 */
export const useDraftTransactionWithTxParams = () => {
  const draftTransaction = useSelector(getCurrentDraftTransaction);

  const unapprovedTxs = useSelector(getUnapprovedTransactions);

  let transactionData = {};

  if (Object.keys(draftTransaction).length !== 0) {
    const editingTransaction = unapprovedTxs[draftTransaction.id];
    transactionData = {
      txParams: {
        gasPrice: draftTransaction.gas?.gasPrice,
        gas: editingTransaction?.userEditedGasLimit
          ? editingTransaction?.txParams?.gas
          : draftTransaction.gas?.gasLimit,
        maxFeePerGas: editingTransaction?.txParams?.maxFeePerGas
          ? editingTransaction?.txParams?.maxFeePerGas
          : draftTransaction.gas?.maxFeePerGas,
        maxPriorityFeePerGas: editingTransaction?.txParams?.maxPriorityFeePerGas
          ? editingTransaction?.txParams?.maxPriorityFeePerGas
          : draftTransaction.gas?.maxPriorityFeePerGas,
        value: draftTransaction.amount?.value,
        type: draftTransaction.transactionType,
      },
      userFeeLevel: editingTransaction?.userFeeLevel,
    };
  }

  return transactionData;
};
