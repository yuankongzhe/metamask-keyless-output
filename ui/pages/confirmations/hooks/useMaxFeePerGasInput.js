import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import { checkNetworkAndAccountSupports1559 } from '../../../selectors';
import { isLegacyTransaction } from '../../../helpers/utils/transactions.util';

import { hexWEIToDecGWEI } from '../../../../shared/modules/conversion.utils';
import { feeParamsAreCustom, getGasFeeEstimate } from './utils';

const getMaxFeePerGasFromTransaction = (transaction, gasFeeEstimates) => {
  if (gasFeeEstimates?.[transaction?.userFeeLevel]) {
    return gasFeeEstimates[transaction.userFeeLevel].suggestedMaxFeePerGas;
  }
  const { maxFeePerGas, gasPrice } = transaction?.txParams || {};
  return Number(hexWEIToDecGWEI(maxFeePerGas || gasPrice));
};

/**
 * @typedef {object} MaxFeePerGasInputReturnType
 * @property {(DecGweiString) => void} setMaxFeePerGas - state setter method to update the maxFeePerGas.
 * @property {string} [maxFeePerGas] - getter method for maxFeePerGas.
 */

/**
 * @param options
 * @param options.estimateToUse
 * @param options.gasEstimateType
 * @param options.gasFeeEstimates
 * @param options.transaction
 * @returns {MaxFeePerGasInputReturnType}
 */
export function useMaxFeePerGasInput({
  estimateToUse,
  gasEstimateType,
  gasFeeEstimates,
  transaction,
}) {
  const supportsEIP1559 =
    useSelector(checkNetworkAndAccountSupports1559) &&
    !isLegacyTransaction(transaction?.txParams);

  const initialMaxFeePerGas = supportsEIP1559
    ? getMaxFeePerGasFromTransaction(transaction, gasFeeEstimates)
    : 0;

  // This hook keeps track of a few pieces of transitional state. It is
  // transitional because it is only used to modify a transaction in the
  // metamask (background) state tree.
  const [maxFeePerGas, setMaxFeePerGas] = useState(() => {
    if (initialMaxFeePerGas && feeParamsAreCustom(transaction)) {
      return initialMaxFeePerGas;
    }
    return null;
  });

  useEffect(() => {
    if (supportsEIP1559 && initialMaxFeePerGas) {
      setMaxFeePerGas(initialMaxFeePerGas);
    }
  }, [initialMaxFeePerGas, setMaxFeePerGas, supportsEIP1559]);

  // We specify whether to use the estimate value by checking if the state
  // value has been set. The state value is only set by user input and is wiped
  // when the user selects an estimate. Default here is '0' to avoid bignumber
  // errors in later calculations for nullish values.
  const maxFeePerGasToUse =
    maxFeePerGas ??
    getGasFeeEstimate(
      'suggestedMaxFeePerGas',
      gasFeeEstimates,
      gasEstimateType,
      estimateToUse,
      initialMaxFeePerGas || 0,
    );

  return {
    maxFeePerGas: maxFeePerGasToUse,
    setMaxFeePerGas,
  };
}
