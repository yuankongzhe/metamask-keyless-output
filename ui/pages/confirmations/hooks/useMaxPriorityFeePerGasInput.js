import { useSelector } from 'react-redux';
import { useEffect, useState } from 'react';

import { checkNetworkAndAccountSupports1559 } from '../../../selectors';
import { isLegacyTransaction } from '../../../helpers/utils/transactions.util';

import { hexWEIToDecGWEI } from '../../../../shared/modules/conversion.utils';
import { feeParamsAreCustom, getGasFeeEstimate } from './utils';

const isNullOrUndefined = (value) => value === null || value === undefined;

const getMaxPriorityFeePerGasFromTransaction = (
  transaction,
  gasFeeEstimates,
) => {
  if (gasFeeEstimates?.[transaction?.userFeeLevel]) {
    return gasFeeEstimates[transaction.userFeeLevel]
      .suggestedMaxPriorityFeePerGas;
  }
  const { maxPriorityFeePerGas, maxFeePerGas, gasPrice } =
    transaction?.txParams || {};
  const feeInHexWei = maxPriorityFeePerGas || maxFeePerGas || gasPrice;
  return feeInHexWei ? Number(hexWEIToDecGWEI(feeInHexWei)) : null;
};

/**
 * @typedef {object} MaxPriorityFeePerGasInputReturnType
 * @property {DecGweiString} [maxPriorityFeePerGas] - the maxPriorityFeePerGas
 *  input value.
 * @property {(DecGweiString) => void} setMaxPriorityFeePerGas - state setter
 *  method to update the maxPriorityFeePerGas.
 */

/**
 * @param options
 * @param options.estimateToUse
 * @param options.gasEstimateType
 * @param options.gasFeeEstimates
 * @param options.transaction
 * @returns {MaxPriorityFeePerGasInputReturnType}
 */
export function useMaxPriorityFeePerGasInput({
  estimateToUse,
  gasEstimateType,
  gasFeeEstimates,
  transaction,
}) {
  const supportsEIP1559 =
    useSelector(checkNetworkAndAccountSupports1559) &&
    !isLegacyTransaction(transaction?.txParams);

  const initialMaxPriorityFeePerGas = supportsEIP1559
    ? getMaxPriorityFeePerGasFromTransaction(transaction, gasFeeEstimates)
    : null;

  const [maxPriorityFeePerGas, setMaxPriorityFeePerGas] = useState(() => {
    if (
      !isNullOrUndefined(initialMaxPriorityFeePerGas) &&
      feeParamsAreCustom(transaction)
    ) {
      return initialMaxPriorityFeePerGas;
    }
    return null;
  });

  useEffect(() => {
    if (supportsEIP1559 && !isNullOrUndefined(initialMaxPriorityFeePerGas)) {
      setMaxPriorityFeePerGas(initialMaxPriorityFeePerGas);
    }
  }, [initialMaxPriorityFeePerGas, setMaxPriorityFeePerGas, supportsEIP1559]);

  const maxPriorityFeePerGasToUse =
    maxPriorityFeePerGas ??
    getGasFeeEstimate(
      'suggestedMaxPriorityFeePerGas',
      gasFeeEstimates,
      gasEstimateType,
      estimateToUse,
      initialMaxPriorityFeePerGas || 0,
    );

  return {
    maxPriorityFeePerGas: maxPriorityFeePerGasToUse,
    setMaxPriorityFeePerGas,
  };
}
