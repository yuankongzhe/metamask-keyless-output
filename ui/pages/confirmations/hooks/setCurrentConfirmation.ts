import { useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { updateCurrentConfirmation } from '../../../ducks/confirm/confirm';
import useCurrentConfirmation from './useCurrentConfirmation';

/*
 * This hook is called from <Confirm /> component to set current transaction.
 * This hook should be invoked only when we are setting or resetting the
 * current confirmation displayed to the user.
 */
const setCurrentConfirmation = () => {
  const dispatch = useDispatch();
  const { currentConfirmation } = useCurrentConfirmation();

  useEffect(() => {
    if (currentConfirmation) {
      dispatch(updateCurrentConfirmation(currentConfirmation));
    }
  }, [currentConfirmation]);

  useEffect(() => {
    return () => {
      dispatch(updateCurrentConfirmation(undefined));
    };
  }, []);
};

export default setCurrentConfirmation;
