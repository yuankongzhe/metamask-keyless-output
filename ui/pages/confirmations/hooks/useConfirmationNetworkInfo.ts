import { useSelector } from 'react-redux';

import {
  NETWORK_TO_NAME_MAP,
  NETWORK_TYPES,
} from '../../../../shared/constants/network';

import {
  currentConfirmationSelector,
  getAllNetworks,
} from '../../../selectors';
import { getProviderConfig } from '../../../ducks/metamask/metamask';

import { useI18nContext } from '../../../hooks/useI18nContext';

type KeyOfNetworkName = keyof typeof NETWORK_TO_NAME_MAP;

function useConfirmationNetworkInfo() {
  const currentConfirmation = useSelector(currentConfirmationSelector);
  const allNetworks = useSelector(getAllNetworks);
  const providerConfig = useSelector(getProviderConfig);
  const t = useI18nContext();

  let networkDisplayName = '';
  let confirmationNetwork;

  if (currentConfirmation) {
    // use the current confirmation chainId, else use the current network chainId
    const currentChainId =
      currentConfirmation?.chainId ?? providerConfig.chainId;
    confirmationNetwork = allNetworks.find(
      ({ chainId }) => chainId === currentChainId,
    );
    if (confirmationNetwork) {
      const { nickname, providerType, type } = confirmationNetwork;
      if (providerType === NETWORK_TYPES.RPC || type === NETWORK_TYPES.RPC) {
        networkDisplayName = nickname ?? t('privateNetwork');
      } else {
        networkDisplayName =
          NETWORK_TO_NAME_MAP[confirmationNetwork?.chainId as KeyOfNetworkName];
      }
    }
  }

  return {
    networkImageUrl: confirmationNetwork?.rpcPrefs?.imageUrl ?? '',
    networkDisplayName,
  };
}

export default useConfirmationNetworkInfo;
