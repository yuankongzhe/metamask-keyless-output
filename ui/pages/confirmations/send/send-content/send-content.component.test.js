import React from 'react';
import { act, waitFor } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';

import { renderWithProvider } from '../../../../../test/lib/render-helpers';
import mockSendState from '../../../../../test/data/mock-send-state.json';
import {
  NETWORK_TYPES,
  CHAIN_IDS,
  GOERLI_DISPLAY_NAME,
} from '../../../../../shared/constants/network';
import { useIsOriginalNativeTokenSymbol } from '../../../../hooks/useIsOriginalNativeTokenSymbol';
import SendContent from '.';

jest.mock('../../../../store/actions', () => ({
  gasFeeStartPollingByNetworkClientId: jest
    .fn()
    .mockResolvedValue('pollingToken'),
  gasFeeStopPollingByPollingToken: jest.fn(),
  getNetworkConfigurationByNetworkClientId: jest
    .fn()
    .mockResolvedValue({ chainId: '0x5' }),
  createTransactionEventFragment: jest.fn(),
  getGasFeeTimeEstimate: jest.fn().mockResolvedValue('unknown'),
  getTokenSymbol: jest.fn().mockResolvedValue('ETH'),
}));

jest.mock('../../../../hooks/useIsOriginalNativeTokenSymbol', () => {
  return {
    useIsOriginalNativeTokenSymbol: jest.fn(),
  };
});

const render = async (props, overrideStoreState) => {
  const mockStore = configureMockStore()({
    ...mockSendState,
    metamask: {
      ...mockSendState.metamask,
      providerConfig: {
        chainId: CHAIN_IDS.GOERLI,
        nickname: GOERLI_DISPLAY_NAME,
        type: NETWORK_TYPES.GOERLI,
      },
    },
    ...overrideStoreState,
  });

  let result;

  await act(
    async () =>
      (result = renderWithProvider(<SendContent {...props} />, mockStore)),
  );

  return result;
};

describe('SendContent Component', () => {
  beforeEach(() => {
    useIsOriginalNativeTokenSymbol.mockReturnValue(true);
  });

  describe('render', () => {
    it('should match snapshot', async () => {
      const { container } = await render({
        gasIsExcessive: false,
        showHexData: true,
      });

      await waitFor(() => {
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('SendHexDataRow', () => {
    it('should not render the SendHexDataRow if props.showHexData is false', async () => {
      const { queryByText } = await render({
        gasIsExcessive: false,
        showHexData: false,
      });

      await waitFor(() => {
        expect(queryByText('Hex data:')).not.toBeInTheDocument();
      });
    });

    it('should not render the SendHexDataRow if the asset type is TOKEN (ERC-20)', async () => {
      const tokenAssetState = {
        send: {
          ...mockSendState.send,
          draftTransactions: {
            '1-tx': {
              ...mockSendState.send.draftTransactions['1-tx'],
              asset: {
                balance: '0x3635c9adc5dea00000',
                details: {
                  address: '0xAddress',
                  decimals: 18,
                  symbol: 'TST',
                  balance: '1',
                  standard: 'ERC20',
                },
                error: null,
                type: 'TOKEN',
              },
            },
          },
        },
      };

      const { queryByText } = await render(
        {
          gasIsExcessive: false,
          showHexData: true,
        },
        tokenAssetState,
      );

      await waitFor(() => {
        expect(queryByText('Hex data:')).not.toBeInTheDocument();
      });
    });
  });

  describe('Gas Error', () => {
    it('should show gas warning when gasIsExcessive prop is true.', async () => {
      const { queryByTestId } = await render({
        gasIsExcessive: true,
        showHexData: false,
      });

      const gasWarning = queryByTestId('gas-warning-message');

      await waitFor(() => {
        expect(gasWarning).toBeInTheDocument();
      });
    });

    it('should show gas warning for none gasEstimateType in state', async () => {
      const noGasPriceState = {
        metamask: {
          ...mockSendState.metamask,
          gasEstimateType: 'none',
          providerConfig: {
            chainId: CHAIN_IDS.GOERLI,
            nickname: GOERLI_DISPLAY_NAME,
            type: NETWORK_TYPES.GOERLI,
          },
        },
      };

      const { queryByTestId } = await render(
        {
          gasIsExcessive: false,
          showHexData: false,
        },
        noGasPriceState,
      );

      const gasWarning = queryByTestId('gas-warning-message');

      await waitFor(() => {
        expect(gasWarning).toBeInTheDocument();
      });
    });
  });

  describe('Recipient Warning', () => {
    it('should show recipient warning with knownAddressRecipient state in draft transaction state', async () => {
      const knownRecipientWarningState = {
        send: {
          ...mockSendState.send,
          draftTransactions: {
            '1-tx': {
              ...mockSendState.send.draftTransactions['1-tx'],
              recipient: {
                ...mockSendState.send.draftTransactions['1-tx'].recipient,
                warning: 'knownAddressRecipient',
              },
            },
          },
        },
        metamask: {
          ...mockSendState.metamask,
          gasEstimateType: 'none',
          providerConfig: {
            chainId: CHAIN_IDS.GOERLI,
            nickname: GOERLI_DISPLAY_NAME,
            type: NETWORK_TYPES.GOERLI,
          },
        },
      };

      const { queryByTestId } = await render(
        {
          gasIsExcessive: false,
          showHexData: false,
        },
        knownRecipientWarningState,
      );

      const sendWarning = queryByTestId('send-warning');

      await waitFor(() => {
        expect(sendWarning).toBeInTheDocument();
      });
    });
  });

  describe('Assert Error', () => {
    it('should render dialog error with asset error in draft transaction state', async () => {
      const assertErrorState = {
        send: {
          ...mockSendState.send,
          draftTransactions: {
            '1-tx': {
              ...mockSendState.send.draftTransactions['1-tx'],
              asset: {
                ...mockSendState.send.draftTransactions['1-tx'].asset,
                error: 'transactionError',
              },
            },
          },
        },
        metamask: {
          ...mockSendState.metamask,
          gasEstimateType: 'none',
          providerConfig: {
            chainId: CHAIN_IDS.GOERLI,
            nickname: GOERLI_DISPLAY_NAME,
            type: NETWORK_TYPES.GOERLI,
          },
        },
      };

      const { queryByTestId } = await render(
        {
          gasIsExcessive: false,
          showHexData: false,
        },
        assertErrorState,
      );

      const dialogMessage = queryByTestId('dialog-message');

      await waitFor(() => {
        expect(dialogMessage).toBeInTheDocument();
      });
    });
  });

  describe('Warning', () => {
    it('should display warning dialog message from warning prop', async () => {
      const { queryByTestId } = await render({
        gasIsExcessive: false,
        showHexData: false,
        warning: 'warning',
      });

      const dialogMessage = queryByTestId('dialog-message');

      await waitFor(() => {
        expect(dialogMessage).toBeInTheDocument();
      });
    });
  });
});
