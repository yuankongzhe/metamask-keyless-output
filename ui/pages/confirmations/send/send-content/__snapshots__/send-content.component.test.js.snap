// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendContent Component render should match snapshot 1`] = `
<div>
  <div
    class="page-container__content"
  >
    <div
      class="send-v2__form"
    >
      <div
        class="send-v2__form-row"
      >
        <div
          class="send-v2__form-label"
        >
          Asset:
        </div>
        <div
          class="send-v2__form-field"
        >
          <div
            class="send-v2__asset-dropdown"
          >
            <div
              class="send-v2__asset-dropdown__input-wrapper"
            >
              <div
                class="send-v2__asset-dropdown__asset"
              >
                <div
                  class="send-v2__asset-dropdown__asset-icon"
                >
                  <div
                    class="identicon__image-border"
                    style="height: 36px; width: 36px; border-radius: 18px;"
                  />
                </div>
                <div
                  class="send-v2__asset-dropdown__asset-data"
                >
                  <div
                    class="send-v2__asset-dropdown__symbol"
                  />
                  <div
                    class="send-v2__asset-dropdown__name"
                  >
                    <span
                      class="send-v2__asset-dropdown__name__label"
                    >
                      Balance:
                    </span>
                    <div
                      class="mm-box currency-display-component mm-box--display-flex mm-box--flex-wrap-wrap mm-box--align-items-center"
                      title="0 ETH"
                    >
                      <span
                        class="mm-box mm-text currency-display-component__text mm-text--inherit mm-text--ellipsis mm-box--color-text-default"
                      >
                        0
                      </span>
                      <span
                        class="mm-box mm-text currency-display-component__suffix mm-text--inherit mm-box--margin-inline-start-1 mm-box--color-text-default"
                      >
                        ETH
                      </span>
                    </div>
                  </div>
                </div>
                <i
                  class="fa fa-caret-down fa-lg send-v2__asset-dropdown__caret"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="send-v2__form-row"
      >
        <div
          class="send-v2__form-label"
        >
          Amount:
          <button
            class="send-v2__amount-max"
          >
            <input
              readonly=""
              type="checkbox"
            />
            <div
              class="send-v2__amount-max__button"
            >
              Max
            </div>
          </button>
        </div>
        <div
          class="send-v2__form-field-container"
        >
          <div
            class="send-v2__form-field"
          >
            <div
              class="unit-input"
            >
              <div
                class="unit-input__inputs"
              >
                <div
                  class="unit-input__input-container"
                >
                  <input
                    class="unit-input__input"
                    data-testid="currency-input"
                    dir="ltr"
                    placeholder="0"
                    style="width: 1.5ch;"
                    type="number"
                    value="1"
                  />
                  <div
                    class="unit-input__suffix"
                  >
                    ETH
                  </div>
                </div>
                <div
                  class="currency-input__conversion-component"
                >
                  No conversion rate available
                </div>
              </div>
              <button
                class="currency-input__swap-component"
                data-testid="currency-swap"
              >
                <i
                  class="fa fa-retweet fa-lg"
                />
              </button>
            </div>
          </div>
          <div />
        </div>
      </div>
      <div
        class="send-v2__form-row"
      >
        <div
          class="send-v2__form-label"
        >
          Hex data:
        </div>
        <div
          class="send-v2__form-field-container"
        >
          <div
            class="send-v2__form-field"
          >
            <textarea
              class="send-v2__hex-data__input"
              placeholder="Optional"
            />
          </div>
          <div />
        </div>
      </div>
      <div
        class="box gas-display box--flex-direction-row"
      >
        <div
          class="transaction-detail"
        >
          <div
            class="transaction-detail-rows"
          >
            <div
              class="transaction-detail-item"
              data-testid="confirm-gas-display"
            >
              <div
                class="transaction-detail-item__row"
              >
                <h6
                  class="mm-box mm-text mm-text--body-md mm-text--font-weight-bold mm-box--display-flex mm-box--flex-wrap-nowrap mm-box--align-items-center mm-box--color-text-default"
                >
                  <p
                    class="mm-box mm-text mm-text--body-md mm-box--color-text-default"
                  >
                    Estimated fee
                  </p>
                </h6>
                <div
                  class="transaction-detail-item__detail-values"
                >
                  <h6
                    class="mm-box mm-text mm-text--body-md mm-text--font-weight-bold mm-text--text-align-right mm-box--margin-left-1 mm-box--color-text-default"
                  >
                    <div
                      class="gas-details-item__currency-container"
                    >
                      <div
                        class="mm-box currency-display-component mm-box--display-flex mm-box--flex-wrap-wrap mm-box--align-items-center"
                        title="0.0000315 ETH"
                      >
                        <span
                          class="mm-box mm-text currency-display-component__text mm-text--inherit mm-text--ellipsis mm-box--color-text-default"
                        >
                          0.0000315
                        </span>
                        <span
                          class="mm-box mm-text currency-display-component__suffix mm-text--inherit mm-box--margin-inline-start-1 mm-box--color-text-default"
                        >
                          ETH
                        </span>
                      </div>
                    </div>
                  </h6>
                </div>
              </div>
              <div
                class="transaction-detail-item__row"
              >
                <div>
                  <div
                    class="mm-box mm-box--display-flex"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-xs mm-box--color-text-muted"
                    >
                      Unknown processing time
                    </p>
                    <p
                      class="mm-box mm-text mm-text--body-xs mm-box--margin-left-1 mm-box--color-error-default"
                    >
                      <span
                        data-testid="gas-timing-time"
                      >
                        ~
                        
                      </span>
                    </p>
                  </div>
                </div>
                <h6
                  class="mm-box mm-text transaction-detail-item__row-subText mm-text--body-sm mm-text--text-align-end mm-box--color-text-alternative"
                >
                  <div
                    class="box gas-details-item__gasfee-label box--display-inline-flex box--flex-direction-row"
                  >
                    <div
                      class="box box--margin-right-1 box--flex-direction-row"
                    >
                      <strong>
                        Max fee:
                      </strong>
                    </div>
                    <div
                      class="gas-details-item__currency-container"
                    >
                      <div
                        class="mm-box currency-display-component mm-box--display-flex mm-box--flex-wrap-wrap mm-box--align-items-center"
                        title="0.0000315 ETH"
                      >
                        <span
                          class="mm-box mm-text currency-display-component__text mm-text--inherit mm-text--ellipsis mm-box--color-text-default"
                        >
                          0.0000315
                        </span>
                        <span
                          class="mm-box mm-text currency-display-component__suffix mm-text--inherit mm-box--margin-inline-start-1 mm-box--color-text-default"
                        >
                          ETH
                        </span>
                      </div>
                    </div>
                  </div>
                </h6>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
